java/ql/src/semmle/code/java/dataflow/ExternalFlow.qll

```diff
@@ -470,11 +471,29 @@ private string getLast(string s) {
 }
 
 private predicate parseParam(string c, int n) {
-  specSplit(_, c, _) and c.regexpCapture("Parameter\\[([-0-9]+)\\]", 1).toInt() = n
+  specSplit(_, c, _) and
+  (
+    c.regexpCapture("Parameter\\[([-0-9]+)\\]", 1).toInt() = n
+    or
+    exists(int n1, int n2 |
+      c.regexpCapture("Parameter\\[([-0-9]+)\\.\\.([0-9]+)\\]", 1).toInt() = n1 and
+      c.regexpCapture("Parameter\\[([-0-9]+)\\.\\.([0-9]+)\\]", 2).toInt() = n2 and
+      n = [n1 .. n2]
+    )
+  )
 }
 
 private predicate parseArg(string c, int n) {
-  specSplit(_, c, _) and c.regexpCapture("Argument\\[([-0-9]+)\\]", 1).toInt() = n
+  specSplit(_, c, _) and
+  (
+    c.regexpCapture("Argument\\[([-0-9]+)\\]", 1).toInt() = n
+    or
+    exists(int n1, int n2 |
+      c.regexpCapture("Argument\\[([-0-9]+)\\.\\.([0-9]+)\\]", 1).toInt() = n1 and
+      c.regexpCapture("Argument\\[([-0-9]+)\\.\\.([0-9]+)\\]", 2).toInt() = n2 and
+      n = [n1 .. n2]
+    )
```

Comment:
Please pardon my ignorance, but are ranges of argument indices really so common that they need their own syntax? Would it not make more sense to allow, say, a comma-separated list inside the brackets (e.g. `Argument[1,2,3]` instead of `Argument[1..3]`). That way you could specify non-contiguous subsets as well, and if you change the corresponding regex to match `[-0-9]+` surrounded by word boundaries then you might even be able to collapse it into a single regex. (It's slightly awkward, since hyphen counts as a word boundary, but something like `(-?\b\d+\b)` should do the trick, I think.)

Also, `[-0-9]+` will match things like `--3-4--4` which might not be desirable, I guess?

Reply:
Using comma-separated lists instead of ranges is a slight generalisation that I don't know whether is needed. It's true that it might be captured by a single regex, but that regex would be somewhat more complex than these two. In the end I guess it comes down to whether ranges or lists are considered nicer UI.

Matching `--3-4--4` doesn't matter - it'll just be filtered by the `.toInt()` rather than the regex, so this way keeps the regex nice and simple.